### ISSUE TITLE
How does batch normalization work in input

### ISSUE THREAD

**Issue by TDDFT** [tensorflow/ranking#42]
There is an example with batch normalization. But I am still confused about the process. 

First, the input tensor has the shape of [list size, num features]. Does different normalization apply to individual elements in the matrix of [list size, num features]? I think we should actually do normalization on different feature columns but not rows.

Second, there are padding elements filled with zeros. It looks like they will contribute to the normalization process as well. This seems to be very fishy. Can someone confirm?

**Comment by eggie5** [comment]
I don't think batch norm has anything to do w/ the TF Ranking module per se. Batch norm will only be applicable in your scoring function, which you can consult any of the literature on BN to understand. 

For example, here is a scoring function w/ BN, dropout and an embedding:

```

from tensorflow.feature_column import categorical_column_with_identity, embedding_column


# TODO replace this w/ an estimator
def make_score_fn():
  """Returns a scoring function to build `EstimatorSpec`."""

  def _score_fn(context_features, group_features, mode, params, config):
    """Defines the network to score a documents."""

    del params
    del config
    
    BUCKETS = 900000
    K=3
    print(group_features)
    rest_id = categorical_column_with_identity(key='rid', num_buckets=BUCKETS, default_value=0)
    rest_emb = embedding_column(rest_id, K)
    
    dense_tensor = tf.feature_column.input_layer(group_features, rest_emb)
    
    input_layer = dense_tensor

    is_training = (mode == tf.estimator.ModeKeys.TRAIN)
    
    cur_layer = input_layer
    for i, layer_width in enumerate(int(d) for d in _HIDDEN_LAYER_DIMS):
      cur_layer = tf.layers.dense(cur_layer, units=layer_width)
      cur_layer = tf.layers.batch_normalization(cur_layer, training=is_training)
      cur_layer = tf.nn.relu(cur_layer)
      cur_layer = tf.layers.dropout(cur_layer, rate=.5, training=is_training)

    logits = tf.layers.dense(cur_layer, units=1) #regression
    return logits

  return _score_fn
```

**Comment by xuanhuiwang** [comment]
BN is applied per column in the input layer that has a shape [batch_size, #column].

It is good question about padding. Actually, in score_fn, all we see is real example in tf-ranking. Say, we have list_size = 4. For a list that have 4 examples, we have
[ex1, ex2, ex3, ex4] 
For a list that we have only 2 examples. Internally, we do the following expansion.
[ex1, ex2, ex1, ex2] 

Thus, all the data we see in group_score_fn are real data.

**Comment by TDDFT** [comment]
Thanks @xuanhuiwang for explaining. I am clear about the first question. For the second question, I am reading through the code to try to understand the behavior better. But with your explanation, it could introduce unnecessary bias to the normalization. Hopefully this won't affect the results too much though.

**Comment by xuanhuiwang** [comment]
@TDDFT, yes, the current way is not exact and slightly biased, but seems OK given that BN is sort of approximation anyway.
